next-app/src/routes/password/change/+page.server.ts

```diff
@@ -0,0 +1,33 @@
+import { sf, type Rpc } from '$lib/server/sf'
+import { fetch_current_user } from '$lib/server/user'
+import { fail, redirect } from '@sveltejs/kit'
+import type { RequestEvent } from '../$types'
+
+export const actions = {
+  default: async ({ request }: RequestEvent) => {
+	const submissions = await request.formData()
+
+	const new_password = submissions.get('new_password')
+	const new_password_confirm = submissions.get('new_password_confirm')
+
+	if (! new_password) {
+		return fail(400, { failed: 'Password is required' })
+	}
+
+	if (new_password !== new_password_confirm) {
+		return fail(400, { failed: 'Passwords do not match' })
+	}
+
+	const cookie = request.headers.get('cookie') || ''
+
+	const { id } = await fetch_current_user(cookie)
+
+	await sf({
+		name: 'change_password',
+		args: [id, new_password],
+		cookie,
+	})
+
+	throw redirect(303, '/password/changed')
```

Comment:
`throw` is part of the natural flow? 😲

Reply:
I also find this concerning, but apparently this is how you do it in SvelteKit.  See https://kit.svelte.dev/docs/load#redirects

Reply:
yeah I almost commented on it. It is weird, but it does mean that you can redirect from deep inside some code that might be checking something. At least we're on v1 now so we don't have to worry about this kind of thing changing.

Reply:
So I am still bothered by `throw` being part of the normal code path here.  So I think here's what's going on:
SvelteKit's `redirect()` is treated like `error()` and so you are supposed to use `throw`.  I think the logic goes something like "redirects are an unexpected error state where the user asks for one thing, but it's moved and so you have to go somewhere else (that you didn't expect) to get it"

I wonder if instead it should use:
`import { goto } from '$app/navigation';`
then
`goto('/password/changed');`

Reply:
yes, it's meant to follow the paradigm of "breaking" normal code flow just like a redirect does.  Similar to the way Svelte uses a valid construct to achieve new concepts, e.g., `export let var` and `$:`.  I'm ok with the mental model shift.

Reply:
ok, so it is a mental model shift - I can handle that.

Would you agree that we should `goto` /password/changed for the success message instead of `redirect`ing there?  It seems more natural to goto with a 200 instead of redirecting with a 303 See Other.

Reply:
> I wonder if instead it should use:
> `import { goto } from '$app/navigation';`
> then
> `goto('/password/changed');`

that would probably accomplish the same thing but not necessarily follow the HTTP flow as well and I sort of like the throw in this case because there are many times in the past where I've had to setup a redirect response and then explicitly return the response to early-exit the current code or function.  This approach felt very natural to me the first time I saw it because setting up a response and then return it later always felt like a leak to me when it should just be finished at that point.

Reply:
> ok, so it is a mental model shift - I can handle that.
> 
> Would you agree that we should `goto` /password/changed for the success message instead of `redirect`ing there? It seems more natural to goto with a 200 instead of redirecting with a 303 See Other.

I'd say the 303 is tailer-made for this scenario (https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/303), where the resource in this case was `/password/change` and the new one is `passowrd/changed`.

Reply:
From the MDN docs, it would seem that 302 is appropriate for a success confirmation page.  I have never seen a 302 used in this way before, but a lot of this is new to me so I'll continue to learn.